### PR TITLE
Remove Data Sharing

### PR DESCRIPTION
**Pull request by GaneshSPatil**
Issue: #8271

Description:

* Remove data sharing spa and models.
* Remove data sharing notification.
* Remove usage data reporter.
* Remove data sharing settings API.
* Remove usage data reporting API.
* Remove data sharing from admin dropdown.
* Remove passing data sharing related server id for the update check.
* Cleanup DataSharingSettings and UsageStatisticsReporting tables on an upgrade.
* Cleanup local storage for data sharing related data

### CODE REVIEW

**Comment by GaneshSPatil** [comment]
@maheshp  @arvindsv @kritika-singh3 @marques-work 

Some questions related to this PR:

1. We cleanup the data sharing related tables on an upgrade, should we also cleanup local storage? 

2. We never mentioned anything about data sharing on our docs, so is release notes a good enough place to mention we do not collect data anymore? 
Note: [This](https://www.gocd.org/2018/07/24/gocd-usage-analytics/) blog post was published when we started the data sharing (also updated once, when there was a change in the items that we collect), we can possibly update this blog post to add a note about removal of data sharing. 

3. [Edge Case] Stale data sharing notification:
	Imagine a user installs GoCD v20.5.0 and on server startup receives the notification `GoCD shares data so that it can be improved`. The user does not read the notification (neither closes or clicks on the message link). Now, the user upgrades the server to GoCD v20.6.0, the user will see the same notification even on GoCD v20.6.0, which would be wrong as GoCD does not have data sharing as part of v20.6.0.

**Comment by arvindsv** [comment]
1. I think yes, we should clean up local storage too.

2. Yes, I think a quick mention in release notes and updating the blog post should be fine.

3. :) I think it's fine.

**Comment by maheshp** [comment]
@GaneshSPatil are you also planning to remove the lambda which captures the data.  Once it's removed the users using the older instance of GoCD will see errors in the browser console right until they upgrade to the latest version.

**Comment by marques-work** [comment]
> @GaneshSPatil are you also planning to remove the lambda which captures the data. Once it's removed the users using the older instance of GoCD will see errors in the browser console right until they upgrade to the latest version.

If it's a problem, we could leave it for a while until the lambda requests drop to near-zero? The free tier would probably even cover us just fine. Or we could ignore it and see if anyone asks about the error.

**Comment by GaneshSPatil** [comment]
> > @GaneshSPatil are you also planning to remove the lambda which captures the data. Once it's removed the users using the older instance of GoCD will see errors in the browser console right until they upgrade to the latest version.
> 
> If it's a problem, we could leave it for a while until the lambda requests drop to near-zero? The free tier would probably even cover us just fine. Or we could ignore it and see if anyone asks about the error.

@maheshp @marques-work - I would prefer removing the lambda, because:

1. I don't know what is the good time to wait till the count reaches zero (or near-zero). There are always some users who might not upgrade due to several reasons.

2. I don't know how are ~we~(who is) going to use this data.

3. Also, if we remove it immediately, we will loose out on newly created instance information, the existing instances would have pinged us until now.

> Once it's removed the users using the older instance of GoCD will see errors in the browser console right until they upgrade to the latest version.

@maheshp - there might not be an error on console. Looking at [this](https://github.com/gocd/gocd/blob/master/server/src/main/webapp/WEB-INF/rails/webpack/models/shared/usage_data_reporter.ts#L58-L69) code, it looks like when one of the API call fails, we just mark reporting check done and move on. Though, there would be a failed XHR call on the network tab.

**Comment by maheshp** [comment]
Thanks @GaneshSPatil I would prefer removing the lambda as well.

**Comment by arvindsv** [comment]
I think it would be fine to keep a very small lambda which just throws away its input, so that the older instances continue to be ok, while we remove the code from newer instances (as done here). At least for now. We always have the option of deleting that no-op lambda whenever we want.

**Comment by GaneshSPatil** [comment]
> I think it would be fine to keep a very small lambda which just throws away its input, so that the older instances continue to be ok

AFAIK, other functionality of the GoCD application should not be affected. Is there anything that we're concerned about if the data sharing API call fails?

@arvindsv - the no-op lambda needs to have all the 3 endpoints for everything to work: 1) _to serve certificates_, 2) _to accept data on v1 endpoint_ and 3) _to accept data on v2 endpoint_. 

Instead of investing time in ~re~implementing no-op lambda, we can keep the existing one around. This might add some more thousand records to the existing database.

If we really really really don't want the data going forward, we can take the backup of the existing data now and redeploy the lambda application creating fresh database, which can be thrown away later. :)

**Comment by arvindsv** [comment]
In that case, maybe it's ok to get rid of it, if all it does it to cause a console log message now and then.

**Comment by arvindsv** [comment]
If not, as you all said, we can keep it around. But, if not now, at a later point it will go away. Then what? Might as well face that now and try and mitigate it. For instance, if older versions of GoCD suddenly stop working if it's not available (unlikely), then we can warn people to upgrade and do a proper deprecation.

File: db-support/db-migration/src/main/resources/db-migration-scripts/initial/create-schema.xml [hunk 1]
```diff
@@ -396,17 +396,6 @@
             <column name="CANCELLEDBY" type="${dataType.caseSensitiveString_255}"/>
         </createTable>
     </changeSet>
-    <changeSet author="gocd(generated)" id="24">
```
[the remark this answers is not in the document]
Comment (GaneshSPatil) [reply db-support/db-migration/src/main/resources/db-migration-scripts/initial/create-schema.xml]: These changes were reverted as part of https://github.com/gocd/gocd/pull/8278/commits/0e4eba7d07acf4b20eaa62cdce2fb83e993f0d60 commit

File: db-support/db-migration/src/main/resources/db-migration-scripts/migrations/2020_06_1_remove_data_sharing_tables.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
+
+  <changeSet id="2020_06_1_remove_data_sharing_settings_table" author="gocd">
```
[the remark this answers is not in the document]
Comment (GaneshSPatil) [reply db-support/db-migration/src/main/resources/db-migration-scripts/migrations/2020_06_1_remove_data_sharing_tables.xml:23]: Make sense, will add a fix.

**Comment by kritika-singh3** [comment]
Verified on ``GoCD Version: 20.6.0 (11971-2fe29190f64d8fb2f6bc1089ab0e743b943ecf30). 
Got `404` when tried to access the url for  Data Sharing SPA, Data Sharing and Reporting APIs.

**Pull request merged**